### PR TITLE
Fixing the chat client

### LINKED ISSUE #54598: The characters "<>/\\" cannot be used in the modular computer chat client.
## Testmerges:
- #54483 
- #54552 

## Reproduction:
1. Acquire a modular computer (tablet, laptop, or console)
2. Download the chat client
3. Open a channel
4. Attempt to post a message in the channel containing one or more of the characters "<>/\\"
5. Notice that nothing is posted.

### PR DESCRIPTION
**Pull request by twilightwanderer**
## About The Pull Request

Corrects the chat restriction on using ASII characters only. ([This PR](https://github.com/Skyrat-SS13/Skyrat-tg/pull/7598) was created for Skyrat-TG, but the collaborator of that repository asked me to adapt this PR for tgstation.)

изображение

## Why It's Good For The Game

This will fix #54598
In the original code, the chat client uses only ASII standard characters, which is very limited in its capabilities. For example, does not allow you to use specialized characters, which would have taken the atmosphere of the old messengers, as well as regional characters. The lack of regional characters complicates the game for non-English-speaking servers. For example, the Russian-speaking player community Space Station 13 The Fluffy Frontier uses the original Skyrat-tg build. And the players of this community almost do not use the chat client to communicate due to the fact that they can not use Cyrillic characters. 

I tried to solve this problem by rewriting some of the text handling functions, namely:
- /proc/reject_bad_message(text) - сhecks messages for emptiness;
- /proc/reject_bad_username(text, max_length = 32) - Checks the entered name for emptiness and length. (32 characters seems optimal to me for a nickname. But if not, it's easy to change.);
- /proc/reject_bad_chatname(text, max_length = 12) - Checks the entered name of the chat for emptiness and the number of characters (I chose 12 characters because a large number expands the panel with chats to the size in which you can not see the main message box). 

## Changelog
:cl:
add: Adding the ability to use a large number of characters in the chat client.
code: In the chat client code, the standard text checking procedures have been replaced by special chat text checking procedures.
/:cl:

### CODE REVIEW

**Comment by maxymax13** [comment]
other locale support is good but it's not really an irc channel if you can just drop emojis in there

**Comment by twilightwanderer** [comment]
> 
> other locale support is good but it's not really an irc channel if you can just drop emojis in there

And what's wrong with using emoji, since almost every player can create his own chat? In any case, it's easy to limit it to a couple of lines in the procedure.

**Comment by Ghommie** [comment]
Don't forget to add "this will fix #54598" to the PR description.

**Comment by twilightwanderer** [comment]
> Don't forget to add "this will fix #54598" to the PR description.

Added as per your request

**Comment by twilightwanderer** [comment]
Should I still hope for some kind of response or is it better to close the PR immediately?

**Comment by LemonInTheDark** [comment]
> 
> 
> Should I still hope for some kind of response or is it better to close the PR immediately?

somewhere between 1 day and a month baby.
feel free to bring it up on discord if your pr gets missed, we get a lot and it can be hard to stay on top.

File: code/__HELPERS/text.dm [hunk 1]
```diff
@@ -953,3 +953,81 @@ GLOBAL_LIST_INIT(binary, list("0","1"))
...
+	if (length(text) > max_length)
+		return
+	else
+		for(var/i = 1, i <= length(text), i += length(char))
+			char = text[i]
+			switch(text2ascii(char))
+				if(0 to 31)
+					return
+				if(32)
+					continue
```
[the remark this answers is not in the document]
Comment (twilightwanderer) [reply code/__HELPERS/text.dm:1028]: @Watermelon914  This construction is designed to keep track of invalid characters. [https://en.wikipedia.org/wiki/ASCII#/media/File:USASCII_code_chart.png](ASCII_Code_Chart) Yes, you can put it in a separate procedure, but there is a problem with this solution. The problem is that you have to take out not a part of the code, which you indicated, but the whole loop. Right now this loop looks the same in all three procedures, but if you need to change it at least in one procedure, your solution will immediately lose its meaning.

[the remark this answers is not in the document]
Comment (Watermelon914) [reply code/__HELPERS/text.dm:1064]: except they there is a lot identical in all 3 functions. This just copy and pasted code. Move the common logic out to its own proc and call them in these procs.

[the remark this answers is not in the document]
Comment (Watermelon914) [reply code/__HELPERS/text.dm:1064]: In fact, you can get rid of the most code duplication by just calling reject_bad_message after your if statement to check if the length of the text is greater than the max length.

**Comment by optimumtact** [comment]
please fix your lint issues.

**Comment by twilightwanderer** [comment]
> please fix your lint issues.

I apologize profusely. At the moment I am very busy with my studies and can not find time to finish it. I will try to fix it in the next few days.

**Comment by Watermelon914** [comment]
your CI is failing

**Comment by LemonInTheDark** [comment]
Why are these all webedits?

**Comment by twilightwanderer** [comment]
> Why are these all webedits?

Because I didn't see them working.

изображение

**Comment by LemonInTheDark** [comment]
I'm sorry I don't quite understand

**Comment by twilightwanderer** [comment]
> I'm sorry I don't quite understand

Oh, I'm sorry. For some reason I thought of html tags. It's just that in the case of tgstation, I find it more convenient to edit the changes through the github editor. If this is critical, I can send changes from my local repository.

**Comment by LemonInTheDark** [comment]
We prefer webedits, especially for changes outside of a few lines. They provide at least a small hope that the code was tested locally, which outside of our qa team (the playerbase) yelling at us is really our only method of quality control.
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#pull-request-process

**Comment by twilightwanderer** [comment]
> We prefer standard commits, and would typically require them for changes outside of a few lines. They provide at least a small hope that the code was tested locally, which outside of our qa team (the playerbase) yelling at us is really our only method of quality control outside of the inspection maintainers do
> https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#pull-request-process

I'll keep that in mind. To confirm that I tested my changes, I will record an animation of the chat.

**Comment by LemonInTheDark** [comment]
It's fine, I trust you. Just keep it in mind for future yeah?

**Comment by twilightwanderer** [comment]
> It's fine, I trust you. Just keep it in mind for future yeah?

Absolutely!

**Comment by twilightwanderer** [comment]
> Finally, trollface ascii in chat client

I'm afraid it won't work. To use ASII graphics, you have to redo the text output. I didn't do that.

**Comment by Watermelon914** [comment]
tested the code?

**Comment by twilightwanderer** [comment]
> tested the code?

Of course

File: code/__HELPERS/text.dm [hunk 2]
```diff
@@ -989,3 +989,27 @@ GLOBAL_LIST_INIT(binary, list("0","1"))
...
+ * This procedure is designed to check the text you type into the chat client. 
+ * It checks for invalid characters and the size of the entered text. 
+ */
+/proc/reject_bad_chattext(text, max_length = 256)
+	var/non_whitespace = FALSE
+	var/char = ""
+	if (length(text) > max_length)
+		return
+	else
+		for(var/i = 1, i <= length(text), i += length(char))
```
Comment (Watermelon914) [inline code/__HELPERS/text.dm:1005]: use `for(var/i in 1 to length(text)`
Alternatively, texts are usually stored as arrays so normally you'd be able to do this too but I'm not 100% sure you can do this in byond, so see if it compiles or not
`for(var/i in text)`

Also, try naming the i variable something more detailed, like index.

Comment (LemonInTheDark) [reply code/__HELPERS/text.dm:1005]: I think that'll break for unicode chars, at least if you want the char and not each byte. Same with length() https://secure.byond.com/docs/ref/#/{notes}/Unicode

Comment (Watermelon914) [reply code/__HELPERS/text.dm:1005]: It can at least be changed to 
`for(var/i in 1 to length(text)` as a micro optimization

Comment (LemonInTheDark) [reply code/__HELPERS/text.dm:1005]: it needs to increase by the size of the read char each step right?

Comment (Watermelon914) [reply code/__HELPERS/text.dm:1005]: I see what you mean

**Pull request merged**